### PR TITLE
fix: unable to run extern-defined function in console

### LINKED ISSUE #1877: panicked at 'Copying extern file'
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

Use `extern "./some-file.js"` pointing to a non-existing file.

### I expected this:

To get a nice fancy compiler error.

### Instead, this happened:

I got a panic:

```
thread '<unnamed>' panicked at 'Copying extern file: Os { code: 44, kind: NotFound, message: "No such file or directory" }', libs/wingc/src/jsify.rs:851:24
stack backtrace:
RuntimeError: unreachable
    at std::panicking::rust_panic_with_hook::h911ba172265603df (wasm://wasm/0096afbe:wasm-function[209]:0xc60a8)
    at core::panicking::panic_fmt::h6958314cbbe045b2 (wasm://wasm/0096afbe:wasm-function[1324]:0x175d98)
    at core::result::unwrap_failed::h5a65fd54aeec1613 (wasm://wasm/0096afbe:wasm-function[1028]:0x167445)
    at wingc::jsify::JSifier::jsify_function::h01a90efa70cfc6da (wasm://wasm/0096afbe:wasm-function[56]:0x6d40c)
    at wingc::jsify::JSifier::jsify_statement::h708589efaa039ff7 (wasm://wasm/0096afbe:wasm-function[19]:0x160e1)
    at wingc_compile.command_export (wasm://wasm/0096afbe:wasm-function[1460]:0x17d0a2)
    at Object.invoke (/Users/eladb/.volta/tools/image/packages/winglang/lib/node_modules/winglang/dist/wingc.js:88:37)
    at compile (/Users/eladb/.volta/tools/image/packages/winglang/lib/node_modules/winglang/dist/commands/compile.js:90:38)
    at async testOne (/Users/eladb/.volta/tools/image/packages/winglang/lib/node_modules/winglang/dist/commands/test.js:15:20)
    at async test (/Users/eladb/.volta/tools/image/packages/winglang/lib/node_modules/winglang/dist/commands/test.js:10:9)

```

### Component:

Compiler

### Wing Version:

_No response_

### Node.js Version:

_No response_

### Platform(s):

_No response_

### Anything else?

_No response_

### LINKED ISSUE #2032: Console cannot simulate extern
Wing Console regression introduced by https://github.com/winglang/wing/pull/1994

The simulator now requires the env var `WING_PROJECT_DIR` to be set for extern module resolution to work. The variable needs to be available in the actual function runtimes.

The variable must be set to the absolute path to the folder containing the original wing file. This is needed because extern can do relative imports, so it needs something to be relative to.

### PR DESCRIPTION
**Pull request by MarkMcCulloh**
Root Issue: The previous enhancement to extern relied on an environment variable `WING_PROJECT_DIR` to have an absolute path to where the .w file is. In running the Wing CLI this is set before doing anything so that's fine. However, when arbitrarily running the Simulator this was a hidden requirement. This is what caused the console to fail to run inflight externs.

Unfortunately, my optimistic quest to either avoid copying extern files or avoid absolute paths has failed. Unless we introduce a bundler to wingc or always run the bundler in preflight, we'll need to have an absolute path for extern files to ensure they're reachable everywhere.

Fixes #2032 (no console change needed)
Fixes #1877

CC @ainvoner

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

**Review by staycoolcall911 — APPROVED** [review]
Thanks for the quick fix Mark.
I added the `pr/e2e full` label - please remove it if it doesn't make sense to you.
Also - looks like some tests are failing on this branch.

File: libs/wingc/src/jsify.rs [hunk 1]
```diff
@@ -873,7 +873,26 @@ impl<'a> JSifier<'a> {
 
 		let body = match &func_def.body {
 			FunctionBody::Statements(scope) => self.jsify_scope(scope, context),
-			FunctionBody::External(external_spec) => format!("return (require(require.resolve(\"{external_spec}\", {{paths: [process.env.WING_PROJECT_DIR]}}))[\"{name}\"])({parameters})")
+			FunctionBody::External(external_spec) => {
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:876]: @MarkMcCulloh I am wondering if this mean that we can avoid the copying of extern files altogether, no? (if I understand correctly we are still doing them).

Comment (MarkMcCulloh) [reply libs/wingc/src/jsify.rs:876]: My last extern PR already avoided copying or creating any sort of wrapper file.

File: libs/wingc/examples/compile.rs [hunk 2]
```diff
@@ -12,8 +12,11 @@ pub fn main() {
 	}
 
 	let source_path = &args[1];
+	let source_path = Path::new(source_path);
 
-	let results = compile(Path::new(source_path), None);
+	std::env::set_var("WING_PROJECT_DIR", source_path.parent().unwrap());
```
Comment (MarkMcCulloh) [inline libs/wingc/examples/compile.rs:17]: Not sure if I'm okay with this. Not having a way to canonicalize paths is really hurting.

The environment variable is currently only set in native test contexts.

**Review by staycoolcall911 — COMMENTED** [review]
I'm fine with these changes to restore `extern` in wing console.

**Review by staycoolcall911 — APPROVED** [review]

File: libs/wingc/src/lib.rs [hunk 3]
```diff
@@ -321,7 +326,11 @@ pub fn compile(source_path: &Path, out_dir: Option<&Path>) -> Result<CompilerOut
 	fs::create_dir_all(out_dir).expect("create output dir");
 
 	let app_name = source_path.file_stem().unwrap().to_str().unwrap();
-	let mut jsifier = JSifier::new(out_dir, app_name, true);
+	let project_dir = absolute_project_root
+		.unwrap_or(source_path.parent().unwrap())
+		.to_path_buf();
```
Comment (Chriscbr) [inline libs/wingc/src/lib.rs:331]: Do we check anywhere that this is an absolute path? (maybe add an assert)

**Comment by Chriscbr** [comment]
@MarkMcCulloh if you have time to add to the description a little bit about what were the reasons the console couldn't run extern-defined code previously, and how the fix fixed it, it would be greatly appreciated. (Just curious if the bug limited to the console, then was it because it's using the CLI a different way than we are testing it, or was it an issue with calling into the SDK's `Simulator` class directly, or if there are any ways the console's code could be changed to fix it, etc.)

File: libs/wingc/src/lib.rs (hunk 3 above)
Comment (Chriscbr) [reply libs/wingc/src/lib.rs:331]: Gotcha. Yeah up to you then 😅

File: libs/wingc/project.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
       "dependsOn": ["lint"],
       "executor": "nx:run-commands",
       "options": {
-        "command": "cargo test",
+        "command": "DEBUG='*' cargo test",
```
[the remark this answers is not in the document]
Comment (MarkMcCulloh) [reply libs/wingc/project.json:10]: I'm kinda on the fence, because it won't show any extra output unless it fails, but I'll remove it for now

**Pull request merged**